File: javascript/014/014.js
"use strict";

const collatz = function(n) {
  if (n <= 1 ) {
    return 1;
  } else {
    if (n % 2 !== 0) {
      return (3 * n) + 1;
    } else {
      return ~~(n / 2);
    }
  }
};

const chainLengths = {};

for (let start = 1; start < 1000000; start += 1) {
  let length = 1;

  let n = start;

  while (n > 1) {
    n = collatz(n);

    length += 1;
  }

  chainLengths[start.toString()] = length;
}

console.log(Object.keys(chainLengths)
  .map((start) => [start, chainLengths[start]])
  .sort((a, b) => a[1] - b[1])
  .slice(-1)[0][0]);

File: python/014/014.py
import operator

def collatz(n):
    if n <= 1:
        return 1
    elif n % 2 == 1:
        return (3 * n) + 1
    else:
        return n / 2

chain_lengths = {}

for start in range(1, 1000001):
    length = 1

    n = start

    while n > 1:
        n = collatz(n)

        length += 1

    chain_lengths[start] = length

print([x[0] for x in sorted(iter(chain_lengths.items()), key = operator.itemgetter(1))][-1:][0])
